docs: add parts 2-4 to the tutorial (#6781)
meltano/meltano: docs/src/_getting-started/part4.md

We're going to do light-weight transformations ("little t or (t)"), so-called **"inline data mappings"** to clean up the data before storing them anywhere. We will use these inline data mappings to hide all emails inside the JSON blob we receive. In the Meltano world, these data mappings are also called a stream maps.

<div class="notification is-success">
    <p>If you're having trouble throughout this tutorial, you can always head over to the <a href="https://meltano.com/slack">Slack channel</a> to get help.</p>
</div>

## Installing the transform-field mapper
To add inline data mappings, we need a new plugin. We're going to use the mapper "transform-field". To add this plugin, use the `meltano add mapper` command:

<div class="termy">
```console
$ meltano add mapper transform-field

Added mapper 'transform-field' to your Meltano project
Variant:        transferwise (default)
Repository:     https://github.com/transferwise/pipelinewise-transform-field
Documentation:  https://docs.meltano.com/concepts/plugins#mappers

Installing mapper 'transform-field'...
---> 100%
Installed mapper 'transform-field'

> To learn more about mapper 'transform-field', visit https://hub.meltano.com/mappers/transform-field

Reviewer: ```suggestion
To learn more about mapper 'transform-field', visit https://docs.meltano.com/concepts/plugins#mappers
```

---
title: Part 4 - Transformations, E(t)LT
description: Part 4 - If you're ready to get started with Meltano and run an EL[T] pipeline with a data source and destination of your choosing, you've come to the right place!
layout: doc
weight: 4
---

Let’s learn by example.

Throughout this tutorial, we’ll walk you through the creation of a end-to-end modern E(t)LT stack.

In the previous parts, we set up a E(t)L pipeline, extracting data from GitHub, transforming it inline, and loading it into a local PostgreSQL database.

In this part, we're going to unleash the data-built-tool [dbt](https://www.getdbt.com/) onto our data to transform it into meaningful information.

<div class="notification is-success">
    <p>If you're having trouble throughout this tutorial, you can always head over to the <a href="https://meltano.com/slack">Slack channel</a> to get help.</p>
</div>

## Install and configure the postgres specific dbt transformer

> Dbt transformers come in flavors. Our flavor is `dbt-postgres`. As usual, you can use the `meltano add` command to add it to your project.

Reviewer: ```suggestion
Dbt uses different [adapters](https://docs.getdbt.com/docs/supported-data-platforms) depending on the database/warehouse/platform you use. Meltano transformers match this pattern; in this case our transformer is `dbt-postgres`. As usual, you can use the `meltano add` command to add it to your project.
```

$ meltano config dbt-postgres set password password
&ensp;&ensp;Transformer 'dbt-postgres' setting 'password' was set in `.env`: 'password'
$ meltano config dbt-postgres set dbname postgres
&ensp;&ensp;Transformer 'dbt-postgres' setting 'dbname' was set in `meltano.yml`: 'postgres'
$ meltano config dbt-postgres set schema analytics
&ensp;&ensp;Transformer 'dbt-postgres' setting 'schema' was set in `meltano.yml`: 'analytics'
```
</div>

## Add our source data
The E(t)L pipeline run already added our source data into the schema `tap_github` as table `commits`. So let's add that to dbt to work with:

```bash
mkdir transform/models/tap_github
```

Add a file called `source.yml` into this directory with the following content:

```yaml
config-version: 2
version: 2
sources:
  - name: tap_github     # the name we want to reference this source by
    schema: tap_github   # the schema the raw data was loaded into
    tables:
      - name: commits
```

## Add a transformed model
Add a file called `authors.sql` to the folder `transform/models/tap_github` with the following contents:

```sql
{% raw %}
{{
  config(
    materialized='table'
  )
}}

> with base as (select *
> from {{ source('tap_github', 'commits') }}) {% endraw %}
> 
> select distinct (commit -> 'author' -> 'name') as authors from base

Reviewer: ```suggestion
WITH base AS (

  SELECT *
  FROM {{ source('tap_github', 'commits') }}
  
) {% endraw %}

SELECT 
  DISTINCT (commit -> 'author' -> 'name') AS authors 
FROM base
```

I'm going to be annoying about structuring SQL https://handbook.meltano.com/data-team/sql-style-guide 😈

## Run the transformation process
To create the actual table, we run the dbt model via `meltano invoke dbt-postgres:run`:

<div class="termy">

```console
$ meltano invoke dbt-postgres:run
2022-09-22T12:30:31.842691Z [info     ] Environment 'dev' is active
12:31:01  Running with dbt=1.1.2
12:31:02  Found 1 model, 0 tests, 0 snapshots, 0 analyses, 167 macros, 0 operations, 0 seed files, 1 source, 0 exposures, 0 metrics
12:31:02
12:31:03  Concurrency: 2 threads (target='dev')
12:31:03
12:31:03  1 of 1 START table model analytics.authors ..................................... [RUN]
---> 100%
12:31:03  1 of 1 OK created table model analytics.authors ................................ [SELECT 2 in 0.59s]
12:31:03
12:31:03  Finished running 1 table model in 1.56s.
12:31:03
12:31:03  Completed successfully
12:31:04
12:31:04  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
#
```

</div>

You can check the data inside the database. There should now be a table inside the `analytics` schema called `authors` populated with data.

## Run the complete pipeline

> To check that everything works together as a pipeline, we clean out once more and run the whole E(t)LT pipeline. Drop the tap_github.commits and the analytics.authors tables, clear the state using `meltano state clear dev:tap-github-to-target-postgres` and run the final pipeline alltogether:

Reviewer: Similar comment as above. It may be reasonable to have a utility which spins up a local postgres database and has some of these commands ready to go 🤔 but that can be in another iteration perhaps.

Author: Probably, but it's a tutorial, it's about teaching. I felt it's ok to teach people about state a bit in part 4, though I agree that it would probably also fine to wrap it away, might make adoption even higher.

[info     ] 12:43:20  Concurrency: 2 threads (target='dev') cmd_type=command name=dbt-postgres stdio=stderr
[info     ] 12:43:20                       cmd_type=command name=dbt-postgres stdio=stderr
[info     ] 12:43:20  1 of 1 START table model analytics.authors ..................................... [RUN] [...]
---> 100%
[info     ] 12:43:21  1 of 1 OK created table model analytics.authors .........[...]
[info     ] 12:43:21                       cmd_type=command name=dbt-postgres stdio=stderr
[info     ] 12:43:21  Finished running 1 table model in 1.34s. cmd_type=command name=dbt-postgres stdio=stderr
[info     ] 12:43:21                       cmd_type=command name=dbt-postgres stdio=stderr
[info     ] 12:43:21  Completed successfully cmd_type=command name=dbt-postgres stdio=stderr
info     ] 12:43:21                       cmd_type=command name=dbt-postgres stdio=stderr
[info     ] 12:43:21  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1 cmd_type=command name=dbt-postgres stdio=stderr
[info     ] Block run completed.           block_type=InvokerCommand err=None set_number=1 success=True
```

</div>

> There we have it, a complete E(t)LT pipeline.

Reviewer: ```suggestion
There we have it, a complete E(t)LT pipeline! Congratulations!
```

Thoughts on having a fun picture of Melty or something like that to celebrate?

Also, you can have them run `meltano dragon` to generate a dragon and celebrate that way 😄 Just thinking of a way to make it a bit more playful.

Reviewer 2: Inline :melty-bouncy: maybe?

## Next Steps

> Next, head over to [Part 5, inside the Getting Started Guide](/getting-started/#add-a-loader-to-send-data-to-a-destination).

Reviewer: ```suggestion
Next, head over to [Part 5, inside the Getting Started Guide](/getting-started/#next-steps).
```
